ROADMAP.md

```diff
@@ -0,0 +1,233 @@
+
+This document outlines the roadmap for integrating linear types in
+GHC. It is laid out in stages which we intend to be successive merge
+points into the GHC tree. The _miscellaneous_ section contains issues
+which are parallel to the stages.
+
+The roadmap contains concrete implementation actions as well as more
+open discussion that needs to be better understood. The _remarks_
+section does not contain action, rather items that will eventually be
+documented.
+
+Stage 1: mvp
+------------
+
+The first iteration implements implements linear types purely in the
+type inference (no modification to core or its linter), and has
+monomorphic multiplicities
+
+- [x] Keep track of linearity in equation binders
+- [x] Constructors are linear by default
+- [x] Wildcard patterns rejected in the linear case
+- [ ] Constructors support unrestricted argument (`Unrestricted`)
+- [ ] Full justification for memory-safety in presence of exceptions
+- [ ] Syntax
+    - [ ] Syntax for multiplicity-annotated arrows. Options:
+        - Mimic Latex and write `a ->_l b` (currently used in this
+          roadmap, but this is no suggestion that this is the one to
+          be favoured)
+        - Edsko de Vries [proposes][edsko-blog] `a ->:l b` (with `l` a
+          multiplicity)
+        - …
+    - [ ] Do we need an ASCII short-hand for linear arrow `a ->_1 b`
+         (Unicode has `⊸` which is the standard notation for the
+         linear arrow)? Many proposals seen so far:
+        - Currently illegal: `-o`
+        - Currently legal: `-.`, `-*`, `-:`
+- [ ] Turn into proper extension
+
+Stage 2: multiplicity polymorphism
+----------------------------------
+
+- [ ] Extend multiplicity expressions with variables (including
+      equality, order, and supremum)
+- [ ] Extend unification with multiplicities (otherwise explicit
+      multiplicity applications are necessary). This probably implies
+      some degree of AC unification.
+- [ ] Allow data types and type classes parametrised with
+      multiplicities
+      - [ ] Probably requires kind for multiplicity arguments.
+
+Stage 3: Core linter
+--------------------
+
+This stage is concerned with checking linearity constraints in
+Core. There is no existing program transformation, to the best of our
+knowledge, which becomes incorrect if linearity constraints are
+ignored. But having the linter work with linearity would help ensure
+that. It would also make it possible to enable further program
+transformation or refine existing analyses, by taking advantage of
+linear types.
+
+- [x] Strictness
+
+  There are some cases where strictness of function might affect
+  typability with linear types in way it does not for unrestricted
+  types.
+
+  ```haskell
+  f :: Unrestricted a -> B -- @f@ has an unrestricted arrow
+
+  u :: Unrestricted a -o B
+  u x = f $ case x of Unrestricted y -> Unrestricted y  -- doesn't typecheck
+  u x = case x of Unrestricted y -> f (Unrestricted y) -- typechecks
+  ```
+
```

Comment:
Probably this is already disallowed because the semantics change when x is bottom.